### PR TITLE
[cxxmodules] Do not generate rdict files if we have a module.

### PR DESCRIPTION
**Pull request by vgvassilev**
## Overview

RDICT files store some useful information (in particular about class offsets) in ROOT files to avoid the potentially expensive call to the interpreter if the information is not the PCH. For example, ROOT's `libGeom` and other third-party code. This is done to circumvent the costly call to `ShowMembers` which will require parsing.

CxxModules diminish that benefit as providing a finer grained control over the information deserialization. We see no observable benefits in generating RDICT files next to our pcm files. 

## Performance results

We have run several times all tests in `roottest/io` and `roottest/meta/MakeProject` folders and compared the results with and without generating RDICT (namely with ROOT master with `-Druntime_cxxmodules=On` with and without this PR). They are available [here](https://docs.google.com/spreadsheets/d/1Y_swct0USyX1DMp0tJOPdZzvmvIOTXOS9s7n08wCb2I/edit?usp=sharing) and [here](https://docs.google.com/spreadsheets/d/1s5LbPmhGBfhimirA0GH87nyEKcrGUfCX80H7-fOrjAU/edit?usp=sharing).

### Methodology

  We have a forwarding `root.exe` which essentially calls `/usr/bin/time -v root.exe $@`. We have processed and stored this information in csv files. We have run in:
  * (no)dict1-cold -- Tests are run immediately after restarting the physical test machine.
  * (no)dict1a-cold -- Tests are run immediately after restarting the physical test machine. Second time.
  * (no)dict2 -- Tests are run once again after running (no)dict1-cold;
  * (no)dict3 -- Tests are run once again after running (no)dict2;
  * (no)dict4 -- Tests are run once again after running (no)dict3;
  * (no)dict5 -- Tests are run once again after running (no)dict4.

The values of each column in (no)dict* are the sum of each value produced after running each test under roottest/io.  The average values are computed in the columns under `dict` and `nodict`. The difference is shown in `diff`.

There are two known test failures wrt to modules whose values are ignored.

The machine specification is `Intel(R) Core(TM) i3-7100 CPU @ 3.90GHz` dual core, 8GB RAM and a **HDD**, run on Ubuntu 18.04.

### Results Interpretation

RDICT mostly intend to reduce the memory footprint during IO. We see no direct benefits in the memory footprint. The average of the max RSS of the sum of the max RSS of every test is higher. There is also a tendency towards faster execution.

### CODE REVIEW

**Comment by vgvassilev** [comment]
Continuation of PR #2912.

**Comment by pcanal** [comment]
> The rdict file does not show performance benefits anymore and we are as good as having only a C++ module.

@vgvassilev What scenario did we test it with to come to this conclusion?  What was the performance gains in the old system on those scenarios and what is the performance difference in the new one? (Memory and run-time)?

**Comment by vgvassilev** [comment]
@yamaguchi1024 and @oshadura can you provide some numbers here?

**Comment by yamaguchi1024** [comment]
@pcanal 
Could you tell us which io test you have in mind which can benefit ROOT from having rdict?

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@yamaguchi1024 You will need an example that has a large fraction of libraries with rdicts (eg. hsimple is almost only using libraries that do not have any rdict).

For example after doing ```ctest -R MakeProject```, you could test:
```
{
   gSystem->Load("roottest/root/meta/MakeProject/cms310/cms310.so");
   auto f = TFile::Open("roottest_src_directory/root/meta/MakeProject/CMSSW_3_1_0_pre11-RelValZTT-default-copy.root");
   TTree *tree = nullptr; f->GetObject("Events", tree);
   for(Long64_t e = 0; e < tree->GetEntriesFast(); ++e) tree->GetEntry(e);
}
```
Or other similar large user libraries.

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by vgvassilev** [comment]
@pcanal, we have some initial tests. Please take a look at the PR description. We will run the MakeProject and update the PR description correspondingly.

File: core/dictgen/src/rootcling_impl.cxx [hunk 1]
```diff
@@ -2841,8 +2841,8 @@ int CheckClassesForInterpreterOnlyDicts(cling::Interpreter &interp,
 ////////////////////////////////////////////////////////////////////////////////
 /// Make up for skipping RegisterModule, now that dictionary parsing
 /// is done and these headers cannot be selected anymore.
-
-int FinalizeStreamerInfoWriting(cling::Interpreter &interp, bool writeEmptyRootPCM=false)
+///
+int MaybeEmitROOTPCM(cling::Interpreter &interp, bool writeEmptyRootPCM=false)
```
Comment (Axel-Naumann) [inline core/dictgen/src/rootcling_impl.cxx]: Why "maybe"?

**Comment by pcanal** [comment]
> @yamaguchi1024 You will need an example that has a large fraction of libraries with rdicts (eg. hsimple is almost only using libraries that do not have any rdict).
> 
> For example after doing `ctest -R MakeProject`, you could test:
> 
> ```
> {
>    gSystem->Load("roottest/root/meta/MakeProject/cms310/cms310.so");
>    auto f = TFile::Open("roottest_src_directory/root/meta/MakeProject/CMSSW_3_1_0_pre11-RelValZTT-default-copy.root");
>    TTree *tree = nullptr; f->GetObject("Events", tree);
>    for(Long64_t e = 0; e < tree->GetEntriesFast(); ++e) tree->GetEntry(e);
> }
> ```
> 
> Or other similar large user libraries.

@vgvassilev @yamaguchi1024 When mentioning the MakeProject in the PR description.  Do you mean the test as it is in the repository or the extent test I proposed (the former  does *not* even load the root-pcm file, i.e. it just generates it).

**Comment by vgvassilev** [comment]
We ran the proposed script as `root.exe -l -q a.c` and we get consistent to the above stated results. Without this pr >D=0K,e=0.94s,F=5,I=152,O=8,K=0K,p=0K,t=0K,M=307076K,S=0.10s,U=0.77s

With the pr: >D=0K,e=0.84s,F=0,I=0,O=0,K=0K,p=0K,t=0K,M=305576K,S=0.07s,U=0.76s

Here important are M and U.

File: core/dictgen/src/rootcling_impl.cxx (hunk 1 above)
Comment (vgvassilev) [reply core/dictgen/src/rootcling_impl.cxx]: Hm... that's probably a leftover of something else. Thanks!

**Comment by pcanal** [comment]
@vgvassilev Looks great.  I need one more point of references (that you already run but I can't tell where it is in the spreadsheet), namely the same number for running just runcms310.C)  thanks.

**Comment by vgvassilev** [comment]
Indeed our script is not very sophisticated. The cms310 is the third row in the data sheets (the thing that requires 82 in MakeProject1-cold).

**Comment by pcanal** [comment]
Fair enough, then reading all the header in memory (and creating all the corresponding TClass) takes 1,649,228 of memory compared to  307,076 for reading one entry with rdict and to 305,576 with modules.  We should run the above test with (master + pch - rdict) to confirm my assumptions.

**Comment by vgvassilev** [comment]
@phsft-bot build just on ROOT-ubuntu16/rtcxxmod with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
[cms310.modified.tar.gz](https://github.com/root-project/root/files/2967051/cms310.modified.tar.gz)
[cms310.read.tar.gz](https://github.com/root-project/root/files/2967052/cms310.read.tar.gz)

**Comment by pcanal** [comment]
In order to prevent the loading of (all) the headers in the readcms310 example, I had to significantly tweak the result of MakeProject where several dictionary were missing (enum's pragma were incorrect, some classes were only forward declared, and having the dictionary for the map required adding operator< in a few places).

**Comment by pcanal** [comment]
See my result at https://docs.google.com/spreadsheets/d/1cKDb99gaNNr9MdSYi1_QQL0gHXhy9m7zmI5voyOK8K8/edit?usp=sharing

where the complete result are in the last worksheet (RDICT numbers).

**Comment by pcanal** [comment]
After the complete run of readcms310.C we get (this is all with the current master)

pch+rdict: master and build of test as is (run MAKEP in cms310 subdirectory to build it
pch+rdict: previsou and mv the rdict:  mv cms310/cms310ProjectDict_rdict.pcm cms310/cms310ProjectDict_rdict.pcm.moved
module+rdict: use runtime_cxxmodules="ON"
pch+rdict: previsou and mv the rdict out of the way

configure line:
```
cmake  -DLLVM_BUILD_TYPE="RelWithDebInfo" -Dcxx17="ON" -Droottest="ON" [-Druntime_cxxmodules="ON"] -Dtest="ON" -Dtmva="OFF" -Dxrootd="OFF" /local2/pcanal/cint_working/rootcling/root-module -G Ninja
```

number is kB

pch+rdict | pch-rdict | module+rdict | module-rdict
------------ | ------------- | ------------ | -------------
413604 | 454132 | 490748  | 534556 

So without modules, avoiding to load the headers (i.e using the rdict) saves **40** MB
Replacing the pch with modules add (as expected as the moment) **77** MB
With modules, avoiding to load pcm content (i.e using the rdict) **still** saves **43** MB.

So it seems to be me that the rdict are still as useful as they were, isn't it? (or am I still missing something?)

Thanks,
Philippe.

**Comment by pcanal** [comment]
New version of the test with actually support testing with module for the 'user' library.
[cms310.read.tar.gz](https://github.com/root-project/root/files/2972133/cms310.read.tar.gz)

**Comment by pcanal** [comment]
After the complete run of readcms310.C we get (this is all with the current master)
pch+rdict: master and build of test as is (run MAKEP in cms310 subdirectory to build it
pch-rdict: use previous but mv the rdict: mv cms310/cms310ProjectDict_rdict.pcm cms310/cms310ProjectDict_rdict.pcm.moved
module+rdict: use runtime_cxxmodules="ON"
module-rdict: use previous but mv the rdict out of the way
configure line:
cmake  -DLLVM_BUILD_TYPE="RelWithDebInfo" -Dcxx17="ON" -Droottest="ON" [-Druntime_cxxmodules="ON"] -Dtest="ON" -Dtmva="OFF" -Dxrootd="OFF" /local2/pcanal/cint_working/rootcling/root-module -G Ninja
In the following table, the numbers are kB. They are as reported by /proc/${pid}/status
phase | what | pch+rdict | pch-rdict | module+rdict | module-rdict ------------ | ------------ || ------------ | ------------- | ------------ | ------------- after reading the entries | VmPeak | 413604 | 454132 | 490748 | 513176 after reading the entries | VmRSS | 140204 | 186148 | 169448 | 195008
For VmRSS
So without modules, avoiding to load the headers (i.e using the rdict) saves 46 MB Replacing the pch with modules add (as expected as the moment) 30 MB With modules, avoiding to load pcm content (i.e using the rdict) still saves 25 MB.
For VmPeak
So without modules, avoiding to load the headers (i.e using the rdict) saves 40 MB Replacing the pch with modules add (as expected as the moment) 77 MB With modules, avoiding to load pcm content (i.e using the rdict) still saves 22 MB.
So it seems to be me that the rdict are still as useful as they were, isn't it? (or am I still missing something?)
Thanks, Philippe.
PS. In run-time, I see an average of 1.20s for module+rdict vs an average of 1.3s for module-rdict (for a 'root.exe -b -l -q' average of .2s) ; and expected this is just an 'initialization' cost when increasing the number of entries read the difference stays around .1s

**Comment by vgvassilev** [comment]
> PS. In run-time, I see an average of 1.20s for module+rdict vs an average of 1.3s for module-rdict (for a 'root.exe -b -l -q' average of .2s) ; and expected this is just an 'initialization' cost when increasing the number of entries read the difference stays around .1s

@pcanal, I'd like to get back to this. Rather than not generating the rdict file we could use it only in the cases when it will improve things. What is the best way to detect if we are in `MakeProject` mode?

**Comment by pcanal** [comment]
>  Rather than not generating the rdict file we could use it only in the cases when it will improve things. 

What are the cases where it doesn't?

> What is the best way to detect if we are in MakeProject mode?

What is special about the "MakeProject mode"?

I don't think we can (really) detect it.  What MakeProject does is generate a set of header files, a linkdef file and a build script.  Except for the fact that there no function besides the constructor and destructor, it is distinguishable from regular user code.   The struct are not even guaranteed to be POD because of inheritance.   In addition, the user could take the initial generation and start developing a full library based on it.   So even heuristic bases on the way MakeProject generates the code would detect that you are "really" in MakeProject mode.

**Comment by vgvassilev** [comment]
> > Rather than not generating the rdict file we could use it only in the cases when it will improve things.
> 
> What are the cases where it doesn't?

See the performance data. It doesn't improve things basically for entire roottest -- on the contrary it shows 2% slowdown.

> 
> > What is the best way to detect if we are in MakeProject mode?
> 
> What is special about the "MakeProject mode"?

I was hoping you to be able to tell. You crafted an example on top of roottest's MakeProject which was able to demonstrate the usefulness of rdicts. None of that so far is obvious to me.

> 
> I don't think we can (really) detect it. What MakeProject does is generate a set of header files, a linkdef file and a build script. Except for the fact that there no function besides the constructor and destructor, it is distinguishable from regular user code. The struct are not even guaranteed to be POD because of inheritance. In addition, the user could take the initial generation and start developing a full library based on it. So even heuristic bases on the way MakeProject generates the code would detect that you are "really" in MakeProject mode.

Sigh... Are you saying that the 'patched' version of MakeProject had many more headers thus read data beyond some threshold which makes rdicts significantly improve the situation?

**Comment by pcanal** [comment]
> I was hoping you to be able to tell. You crafted an example on top of roottest's MakeProject which was able to demonstrate the usefulness of rdicts. None of that so far is obvious to me.

I see.  MakeProject was here just a short-cut to producing a large enough library size.  Technically I could have send you any large experimental framework instead.

**Comment by pcanal** [comment]
> It doesn't improve things basically for entire roottest -- on the contrary it shows 2% slowdown.

I think that if we want to harvest this 2% (by skipping the root-pcm sometimes) we are going to have to understand better what in the "user" code (i.e. the set of headers and functionality) and what in our code makes the difference.

For example, if in the large example, the gain from the root-pcm is thanks to avoiding a large number of data-member offset calculations and large number of AST nodes loading from the module to get the information, and (still for example), if the cost in the other cases (the 2%) is due to loading the TProtoClass but not using them because the example does not do any I/O, then the conclusion would be that the "cost of doing faster I/O is the 2% slowdown and that their would be 2 options (speed-up the module case even more than now or speed-up the root-pcm case when no I/O is done). [Of course in either we would have the challenge of doing so without costing (more?) for the other user cases]

**Comment by vgvassilev** [comment]
It seems that rdict files help when we have significant amount of headers needed for I/O. Can we heuristically determine the threshold size? I think that roottest suffers from performance degradation because the rdict files are almost empty and it does not pay off loading and using them.

**Comment by Axel-Naumann** [comment]
Not everything will be covered by modules, for a long time. Thus we will have to keep the rdict file feature alive for years to come. What's the benefit of optimizing this then -  is it only to accelerate roottest? (I'm trying to understand whether it's important to invest dev hours here vs in other areas, taking as main criteria, as always, relevance to our users.)

**Comment by vgvassilev** [comment]
To clarify, we are talking about the case where a library has a module. Users do not use make project often and roottest claims to cover a good amount of their workflows.

As for the dev hours: they are already invested and we know that only under very specific conditions we need rdicts -- it seems that it depends on the amount of header files we parse. If the number is very high, as in Philippe's example, we should load the rdict -- it slows things by 2%...

**Comment by pcanal** [comment]
I think it might be hard to come up with a proper heuristic.  For example is the threshold "per dictionary" (in which case it might be achiveable) or is it an accumulative cost, i.e. per whole set of library (in which case it is impossible) and is it lightly or highly dependent of the fraction of the generated dicitonary entries and/or header that are actually used? 

> Users do not use make project 

MakeProject is irrelevant here.  It is "only" just to quickly produce an emulation of the "Experiment Framework" type of problem size.  The relevant part is (as you noted) the amount of classes needed for I/O.   One part that example does not emulate is the (high) number of libraries and dictionary.    So *yes* this type of problem is extremely relevant and no it is actually not directly represented in roottest. (and it is unknown which fraction of the roottests are exercising this code path (i.e. not only loading but also using the rootpcms).

Also I suppose there are ways to reduce the up-front cost of the rootpcms (for example delaying their loading until needed)

**Comment by vgvassilev** [comment]
> I think it might be hard to come up with a proper heuristic. For example is the threshold "per dictionary" (in which case it might be achiveable) or is it an accumulative cost, i.e. per whole set of library (in which case it is impossible) and is it lightly or highly dependent of the fraction of the generated dicitonary entries and/or header that are actually used?
> 
> > Users do not use make project
> 
> MakeProject is irrelevant here. It is "only" just to quickly produce an emulation of the "Experiment Framework" type of problem size. The relevant part is (as you noted) the amount of classes needed for I/O. One part that example does not emulate is the (high) number of libraries and dictionary. So _yes_ this type of problem is extremely relevant and no it is actually not directly represented in roottest. (and it is unknown which fraction of the roottests are exercising this code path (i.e. not only loading but also using the rootpcms).

In principle the new implementation due to modules should allow us to easily turn off rootpcms and test in the cmssw context. Then I guess the results should be more reliable.

> 
> Also I suppose there are ways to reduce the up-front cost of the rootpcms (for example delaying their loading until needed)

That's what I am after. Do you have something in particular in mind?

**Comment by pcanal** [comment]
>>   Also I suppose there are ways to reduce the up-front cost of the rootpcms (for example delaying their loading until needed)
> That's what I am after. Do you have something in particular in mind

Currently, the rootpcms file is loaded when the library is loaded.  Instead it could be open whenever there is a call to TClassTable::GetProtoClass for one of its content.  This would require:

  (a) have a (persistent) data structure that list of the available content of the rootpcms.
  (b) extend the TClassTable data structure to record that information
  (c) in TClassTable::GetProtoClass if the TProtoClass is no there but there is rootpcms listed, then open the rootpcms and load.

When opening the rootpcms one should load all the TProtoClasses since opening the file takes time.

**Comment by vgvassilev** [comment]
This pull request has a lot of useful information. The current infrastructure is flexible and can disable rootpcms if required for further performance studies. Let's close this now and come back to it if necessary.

**Pull request closed without merging**